adapt/context.py

```diff
@@ -1,75 +1,152 @@
+"""
+This is to Manage Context of a Conversation
+
+Notes
+-----
+    Comments are subject to evaluation and may not reflect intent.
+    Comments should be updated as code is clearly understood.
+"""
 from six.moves import xrange
 
-__author__ = "seanfitz"
+__author__ = "seanfitz, Art McGee"
 
 
 class ContextManagerFrame(object):
     """
     Manages entities and context for a single frame of conversation.
     Provides simple equality querying.
+
+    Attributes
+    ----------
+    entities : []
+        Entities that belong to ContextManagerFrame
+    metadata : {}
+        metadata to describe context belonging to ContextManagerFrame
     """
+
     def __init__(self, entities=[], metadata={}):
+        """
+        Initialize ContextManagerFrame
+
+        Parameters
+        ----------
+        entities : []
+            List of Entities...
+        metadata : {}
+            metadata to describe context?
+
+        """
         self.entities = entities
         self.metadata = metadata
 
     def metadata_matches(self, query={}):
+        """
+        Returns key matches to metadata
+
+        This will check every key in query for a matching key in metadata
+        returning true if every key is in metadata.  query without keys
+        return false.
+
+        Parameters
+        ----------
+        query : {}
+            metadata for matching
+
+        Returns
+        -------
+        bool
+            True: when key count in query is > 0 and all keys in query in
+                self.metadata
+            False: if key count in query is <= 0 or any key in query not
+                found in self.metadata
+
+        """
         result = len(query.keys()) > 0
         for key in query.keys():
             result = result and query[key] == self.metadata.get(key)
 
         return result
 
     def merge_context(self, tag, metadata):
+        """
+        merge into contextManagerFrame new entity and metadata.
+
+        Appends tag as new entity and adds keys in metadata to keys in
+        self.metadata.
+
+        Parameters
+        ----------
+        tag : string
+            entity to be added to self.entities
+        metadata : {}
+            metadata containes keys to be added to self.metadata
+        """
         self.entities.append(tag)
         for k in metadata.keys():
             if k not in self.metadata:
                 self.metadata[k] = k
+                """TODO: Should this be using the key as the value for key?"""
 
 
 class ContextManager(object):
```

Comment:
Probably worth calling out that this is a reference in-memory implementation. Any context manager that supports this interface will do, and in order to support conversations across multiple devices (or injection of context from user profile), would require a different implementation, possibly even a remote service.